### PR TITLE
add Corner losses

### PR DESCRIPTION
**Pull request by tianyuandu**

### CODE REVIEW

File: mmdet/models/losses/gaussian_focal_loss.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        loss = (pos_loss + neg_loss).sum() / pos_num
+    return loss
+
+
+@LOSSES.register_module()
+class GaussianFocalLoss(nn.Module):
+    """ GaussianFocalLoss is a variant of focal loss.
+
+    Please refer to https://arxiv.org/abs/1808.01244 for more details.
+    Code is modified from https://github.com/princeton-vl/CornerNet.
```
Comment (xvjiarui) [inline mmdet/models/losses/gaussian_focal_loss.py:30]: We may point to the specific file.

File: mmdet/models/losses/ae_loss.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    return pull_loss, push_loss
+
+
+@LOSSES.register_module()
+class AELoss(nn.Module):
+    """ Associative Embedding Loss.
+
+    Please refer to https://arxiv.org/abs/1611.05424 for more details.
+    Code is modified from https://github.com/princeton-vl/CornerNet.
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:50]: We may point to the specific file.

File: mmdet/models/losses/gaussian_focal_loss.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import torch.nn as nn
+
+from ..builder import LOSSES
+
+
+def gaussian_focal_loss(pred, target, weight=None, alpha=2.0, gamma=4.0):
+    eps = 1e-12
+    pos_weights = target.eq(1)
```
Comment (xvjiarui) [inline mmdet/models/losses/gaussian_focal_loss.py:8]: `.eq(1)` is not necessary.

File: mmdet/models/losses/gaussian_focal_loss.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                 gamma=4.0,
+                 reduction='none',
+                 loss_weight=1.0):
+        super(GaussianFocalLoss, self).__init__()
+        self.alpha = alpha
+        self.gamma = gamma
+        self.reduction = reduction
+        self.loss_weight = loss_weight
+
+    def forward(self, pred, target, weight=None):
```
Comment (xvjiarui) [inline mmdet/models/losses/gaussian_focal_loss.py:50]: We may add comments about the `target ` if it is not `0/1` binary.

File: mmdet/models/losses/gaussian_focal_loss.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    pos_num = pos_weights.sum()
+    if pos_num < 1:
+        loss = neg_loss.sum()
+    else:
+        loss = (pos_loss + neg_loss).sum() / pos_num
+    return loss
+
+
+@LOSSES.register_module()
+class GaussianFocalLoss(nn.Module):
```
Comment (xvjiarui) [inline mmdet/models/losses/gaussian_focal_loss.py:26]: The gaussian is encoded in `target`. 
So `GaussianFocalLoss` may not be precise.

File: mmdet/models/losses/gaussian_focal_loss.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    if weight is not None:
+        pos_loss *= weight
+        neg_loss *= weight
+
+    pos_num = pos_weights.sum()
+    if pos_num < 1:
+        loss = neg_loss.sum()
+    else:
+        loss = (pos_loss + neg_loss).sum() / pos_num
```
Comment (xvjiarui) [inline mmdet/models/losses/gaussian_focal_loss.py:21]: We may add comments about why `neg_loss` is not normalized.

File: mmdet/models/losses/ae_loss.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        br_e = br_preds[:, br_y, br_x].view(-1, 1)
+        tl_list.append(tl_e)
+        br_list.append(br_e)
+        me_list.append((tl_e + br_e) / 2.0)
+
+    tl_list = torch.cat(tl_list)
+    br_list = torch.cat(br_list)
+    me_list = torch.cat(me_list)
+
+    # assert tl_list.size() == br_list.size()
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:22]: We may keep the `assert`.

File: mmdet/models/losses/ae_loss.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ..builder import LOSSES
+
+
+def ae_loss_per_image(tl_preds, br_preds, match):
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:8]: Add docstring about this function.

File: mmdet/models/losses/ae_loss.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    tl_list = torch.cat(tl_list)
+    br_list = torch.cat(br_list)
+    me_list = torch.cat(me_list)
+
+    # assert tl_list.size() == br_list.size()
+
+    N, M = tl_list.size()
+
+    if N > 0:
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:26]: Add comments about the conditions.

File: mmdet/models/losses/ae_loss.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        pull_loss = torch.tensor(0).type_as(me_list).requires_grad_()
+
+    margin = 1
+    push_loss = torch.tensor(0).type_as(me_list).requires_grad_()
+
+    # confusion matrix of push loss
+    conf_mat = (me_list.expand((N, N, M)).permute(1, 0, 2) - me_list).sum(-1)
+    conf_mat = (1 - torch.eye(N).type_as(me_list)) * (margin - conf_mat.abs())
+
+    if N > 1:
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:39]: Add comments about the conditions.

File: mmdet/models/losses/ae_loss.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    # assert tl_list.size() == br_list.size()
+
+    N, M = tl_list.size()
+
+    if N > 0:
+        pull_loss = (tl_list - me_list).pow(2) + (br_list - me_list).pow(2)
+        pull_loss = pull_loss.sum() / N
+    else:
+        pull_loss = torch.tensor(0).type_as(me_list).requires_grad_()
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:30]: Please refer to https://github.com/open-mmlab/mmdetection/blob/8951a8a70f36d442d8f5a9ecfe6361bbd6dc2adb/mmdet/models/losses/pisa_loss.py#L149 for 0 loss.

File: mmdet/models/losses/gaussian_focal_loss.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+                        weight=None,
+                        alpha=2.0,
+                        gamma=4.0):
+    eps = 1e-12
+    pos_weights = gaussian_target.eq(1)
+    neg_weights = (1 - gaussian_target).pow(gamma)
+    pos_loss = -(pred + eps).log() * (1 - pred).pow(alpha) * pos_weights
+    neg_loss = -(1 - pred + eps).log() * pred.pow(alpha) * neg_weights
+
+    if weight is not None:
```
Comment (xvjiarui) [inline mmdet/models/losses/gaussian_focal_loss.py:17]: Use https://github.com/open-mmlab/mmdetection/blob/e37867530def6e935881a82f8d8861a1be3d09d9/mmdet/models/losses/utils.py#L26 for `weight`. 
Also, add `reduction`.

File: mmdet/models/losses/ae_loss.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        - no object in image: both pull loss and push loss will be 0.
+        - one object in image: push loss will be 0 and pull loss is computed
+            by the two corner of the only object.
+        - more than one objects in image: pull loss is computed by corner pairs
+            from each object, push loss is computed by each object with all
+            other objects. We use confusion matrix with 0 in diagonal to
+            compute the push loss.
+
+    Args:
+        tl_preds (tensor): embedding feature map of left-top corner.
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:26]: Embedding feature map of left-top corner.

File: mmdet/models/losses/ae_loss.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+        br_preds (tensor): embedding feature map of bottim-right corner.
```
Comment (xvjiarui) [inline mmdet/models/losses/ae_loss.py:27]: Similarly

File: mmdet/models/losses/gaussian_focal_loss.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        super(GaussianFocalLoss, self).__init__()
+        self.alpha = alpha
+        self.gamma = gamma
+        self.reduction = reduction
+        self.loss_weight = loss_weight
+
+    def forward(
+            self,
+            pred,
+            target,  # target is gaussian heatmap, not 0/1 binary target.
```
Comment (hellock) [inline mmdet/models/losses/gaussian_focal_loss.py:47]: Use a standard docstring instead of comments.

File: mmdet/models/losses/ae_loss.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+        conf_mat = conf_weight * (margin - conf_mat.sum(-1).abs())
+
+        if N > 1:  # more than one object in current image
+            push_loss = F.relu(conf_mat).sum() / (N * (N - 1))
+
+    return pull_loss, push_loss
+
+
+@LOSSES.register_module()
+class AELoss(nn.Module):
```
Comment (hellock) [inline mmdet/models/losses/ae_loss.py:70]: Using the full name `AssociativeEmbeddingLoss` seems better.
